File: server/src/main/java/gg/mineral/server/network/packet/play/clientbound/SpawnObjectPacket.java
package gg.mineral.server.network.packet.play.clientbound;

import gg.mineral.api.network.packet.Packet;

import io.netty.buffer.ByteBuf;

public final record SpawnObjectPacket(int entityId, byte type, int x, int y, int z, byte pitch, byte yaw, int data)
        implements Packet.OUTGOING {
    @Override
    public void serialize(ByteBuf os) {
        writeVarInt(os, entityId);
        os.writeByte(type);
        os.writeInt(x);
        os.writeInt(y);
        os.writeInt(z);
        os.writeByte(pitch);
        os.writeByte(yaw);
        os.writeInt(data);
    }

    @Override
    public byte getId() {
        return 0x0E;
    }
}

File: server/src/main/kotlin/gg/mineral/server/network/packet/play/clientbound/SpawnObjectPacket.kt
package gg.mineral.server.network.packet.play.clientbound

import gg.mineral.api.network.packet.Packet
import io.netty.buffer.ByteBuf

@JvmRecord
data class SpawnObjectPacket(
    val entityId: Int,
    val type: Byte,
    val x: Int,
    val y: Int,
    val z: Int,
    val pitch: Byte,
    val yaw: Byte,
    val data: Int
) :
    Packet.Outgoing {
    override fun serialize(os: ByteBuf) {
        os.writeVarInt(entityId)
        os.writeByte(type.toInt())
        os.writeInt(x, y, z)
        os.writeByte(pitch, yaw)
        os.writeInt(data)
    }

    override val id: Byte
        get() = 0x0E
}
